python/lib/core/dmod/core/common/collections/cache.py

```diff
@@ -0,0 +1,1015 @@
+        return "{" + \
+            f'"identifier": {self.__data}, ' \
+            f'"data_hash": {self.__data_hash}, ' \
+            f'"last_accessed": "{self.last_accessed}"' + \
+            "}"
+
+
+# Part of Issue https://github.com/NOAA-OWP/DMOD/issues/434
+#   "Make metric computations asynchronous by location"
+class AccessCache(typing.Generic[HashableType], typing.MutableMapping[str, CacheEntry[HashableType]]):
+    """
+    A base class that implements a caching mechanism that organizes entries based on access time and fires events
+    when adding, removing, accessing, and updating entries
+    """
+    def __init__(
+        self,
+        max_size: int = 100,
+        values: typing.Mapping[str, typing.Union[HashableType, CacheEntry[HashableType], None]] = None,
+        on_addition: typing.Union[CACHE_HANDLER, typing.Sequence[CACHE_HANDLER]] = None,
+        on_removal: typing.Union[CACHE_HANDLER, typing.Sequence[CACHE_HANDLER]] = None,
+        on_access: typing.Union[CACHE_HANDLER, typing.Sequence[CACHE_HANDLER]] = None,
+        on_update: typing.Union[CACHE_HANDLER, typing.Sequence[CACHE_HANDLER]] = None,
+        event_router: EventRouter = None,
+        **kwargs
+    ):
+        """
+        Constructor
+
+        Args:
+            max_size: The maximum number of items that this cache may contain. Non-positive numbers or None will store all data
+            values: Preexisting data to add to the cache
+            on_addition: Handlers to call when an item is added to the cache
+            on_removal: Handlers to call when an item is removed from the cache
+            on_access: Handlers to call when an item in the cache is accessed
+            on_update: Handlers to call when an item in the cache is updated
+            event_router: An EventRouter that will call appropriate functions when events are triggered
+            kwargs: Additional items to store in the cache
+        """
+        if values is None:
+            values = {}
+
+        if on_addition is None:
+            on_addition = list()
+
+        if on_removal is None:
+            on_removal = list()
+
+        if on_access is None:
+            on_access = list()
+
+        if on_update is None:
+            on_update = list()
+
+        self.__event_router = event_router or EventRouter()
+
+        self.__event_router.register_handler(ON_ADDITION_KEY, on_addition)
+        self.__event_router.register_handler(ON_REMOVAL_KEY, on_removal)
+        self.__event_router.register_handler(ON_ACCESS_KEY, on_access)
+        self.__event_router.register_handler(ON_UPDATE_KEY, on_update)
+
+        self.__leftover_tasks: typing.List[typing.Awaitable] = list()
+
+        self.__internal_cache: typing.Dict[str, CacheEntry[HashableType]] = dict()
+        self.__earliest: typing.Optional[CacheEntry[HashableType]] = None
+        self.__max_size = max_size if isinstance(max_size, (int, float)) and max_size > 0 else math.inf
+
+        for key, value in values:
```

Comment:
I think this needs to be `values.items()`?

Reply:
Excellent catch. Side note - I really, really hate that python iterates by keys instead of keys+values for their maps.